update repo to fit the current status of rust development

ivila opened the pull request:
1, use url crate as a replacement of former one
2, upgrade version of rustls_pemfile to make test run 
3, upgrade version of webpki-roots to make test run 
4, add no_std build test of no_std target in prepush script

this PR relate to #10

ivila committed:
update repo to fit the current status of rust development

1, use url crate as a replacement of former one
2, upgrade version of rustls_pemfile to make test run
3, upgrade version of webpki-roots to make test run
4, add no_std build test of no_std target in prepush script

ivila commented:
@bobbobbio Can you spend some time to look into this PR, I change some codes of http_io(see the commit message), and moreover, there are some extra things you might care:
1. the definition of authority is different, the origin one in http_io seems to use host as authority, and in the new codes, I use host_str of servo's Url instead.
2. I remove the Path struct.

bobbobbio commented on examples/connection_reuse.rs:
```diff
@@ -5,16 +5,14 @@ use std::io;
     for path in &["/", "/favicon.ico", "/robots.txt"] {
-        let mut url = url.clone();
-        url.path = path.parse()?;
+        let url = Url::parse(format!("{}{}", host, path).as_str())?;
```
instead of doing this, could you instead parse URL outside of the loop and use `Url::push`. `path` here isn't event a path and actually could just be `Option<String>` to represent adding a component.

I like to avoid doing any string concatenation for path-like things (even if you can prove it can't cause issues in this case) since it can lead to issues like double `/` or missing `/`

ivila replied:
done by switch to url.set_path

bobbobbio commented on src/ssl/mod.rs:
```diff
@@ -1,3 +1,4 @@
+#[allow(unused)]
```
whats going on here, is this fixing some existing warning?

ivila replied:
done by change it to dead_code and add comment

bobbobbio commented on examples/no_std.rs:
```diff
@@ -108,7 +108,7 @@ mod no_std {
         let url: Url = opts.url.parse()?;
 
-        let s = MyFakeTcpStream::connect(url.authority.as_ref(), url.port()?)?;
+        let s = MyFakeTcpStream::connect(url.host_str().unwrap(), url.port_or_known_default().unwrap())?;
```
This `unwrap` here and the other places I'm not okay with. In these examples I want them to be more like you would have a real program, and panic on non-HTTP urls isn't good, we need to instead return an error like you did in the other place, I suspect that you didn't because it can get a bit long, I think we should find a more ergonomic way to check for this.

ivila replied:
done by add HttpUrl struct

bobbobbio commented on src/client.rs:
```diff
@@ -26,7 +26,7 @@
 //! fn main() -> Result<()> {
 //!     let url: Url = "http://www.google.com".parse()?;
-//!     let s = TcpStream::connect((url.authority.as_ref(), url.port()?))?;
+//!     let s = TcpStream::connect((url.host_str().unwrap(), url.port_or_known_default().unwrap()))?;
```
same issue as mentioned above here

bobbobbio commented on src/client.rs:
```diff
@@ -41,11 +41,10 @@
 //!     for path in &["/", "/favicon.ico", "/robots.txt"] {
-//!         let mut url = url.clone();
-//!         url.path = path.parse()?;
+//!         let url = Url::parse(format!("{}{}", host, path).as_str())?;
```
I have the same feedback as above about string pasting

bobbobbio commented on src/client.rs:
```diff
@@ -126,11 +125,11 @@ impl HttpRequestBuilder {
             .map_err(|e| Error::ParseError(e.to_string()))?;
         let mut request = HttpRequest::new(method, url.path());
-        request.add_header("Host", url.authority.clone());
+        request.add_header("Host", url.host_str().unwrap().to_string());
```
same issue with `unwrap`

bobbobbio commented on src/client.rs:
```diff
@@ -226,19 +225,27 @@ impl StreamConnector for std::net::TcpStream {
     fn to_stream_addr(url: Url) -> Result<Self::StreamAddr> {
+        use std::str::FromStr;
+        let Some(host) = url.host_str() else {
+            return Err(crate::error::Error::UrlError(String::from("no host")))
```
to me these errors you are returning are a little confusing, IIUC the reason the host or port would be missing would be that the URL had some scheme that the library doesn't support, I think the correct error would be one saying like "unsupported URL scheme XXX"

bobbobbio replied:
To deal with this scheme issue my first idea would be to make some wrapper struct like `HttpUrl(url::Url)` and its constructor could validate that it only contained HTTP urls. Then we could provide methods to to access things like the `host` which could panic if it didn't exist (because we previously validated that it must) This would also help make the examples stay shorter.
If you don't like that, perhaps then we could just have a function exposed they can call that validates the scheme like `let (host, port) = http_get_host_and_port(&url)?` or just  `ensure_http_url(&url)?`

bobbobbio commented on src/server.rs:
```diff
@@ -59,7 +59,8 @@
 //!     });
 //!
-//!     let url = format!("http://localhost:{}/src/server.rs", port);
+//!
```
what is going on here with this change?

is this an unintentional newline?

bobbobbio commented on src/ssl/rustls.rs:
```diff
@@ -23,20 +18,20 @@ fn read_test_cert(name: &str) -> Result<Vec<u8>> {
-            ta.name_constraints,
+            ta.subject.to_vec(),
+            ta.subject_public_key_info.to_vec(),
+            match ta.name_constraints.as_ref() {
+                Some(v) => Option::Some(v.to_vec()),
+                None => Option::None,
```
I think this can be written shorted like `ta.name_constraints.as_ref().map(|c| c.to_vec())`

bobbobbio reviewed (changes requested):
Thank you so much for contributing. I think you proved that this is a good idea assuming everything builds properly.

My main concern with these changes is that I want the examples to return an error rather than crash when the URL has the wrong scheme

bobbobbio commented:
> @bobbobbio Can you spend some time to look into this PR, I change some codes of http_io(see the commit message), and moreover, there are some extra things you might care:
> 
>     1. the definition of authority is different, the origin one in http_io seems to use host as authority, and in the new codes, I use host_str of servo's Url instead.
This change is good, I think host is correct and authority wasn't
>     2. I remove the Path struct.
It seems like it isn't used anymore, so that should be fine right?

ivila commented:
Thanks a lot for your patience, I will make changes in 2 days.

ivila committed:
PR fix:

1) add HttpUrl struct to avoid using unwrap and to make things simpler
2) in the examples, use old url with url.set_path instead of use format
   macro
3) remove extra lines and add comment for the #[allow(dead_code)]

ivila commented:
> > @bobbobbio Can you spend some time to look into this PR, I change some codes of http_io(see the commit message), and moreover, there are some extra things you might care:
> > ```
> > 1. the definition of authority is different, the origin one in http_io seems to use host as authority, and in the new codes, I use host_str of servo's Url instead.
> > ```
> 
> This change is good, I think host is correct and authority wasn't
> 
> > ```
> > 2. I remove the Path struct.
> > ```
> 
> It seems like it isn't used anymore, so that should be fine right?

Yes, it isn't used in the repo anymore, and if some users depends on it, the can switch to [url.path_segments](https://docs.rs/url/latest/url/struct.Url.html#method.path_segments)
![image](https://github.com/user-attachments/assets/66e64e22-320f-4df2-bbbc-2575d6fc7bbd)

ivila reviewed (commented):
1) add HttpUrl struct to avoid using unwrap and to make things simpler
2) in the examples, use old url with url.set_path instead of use format
   macro
3) remove extra lines and add comment for the #[allow(dead_code)]

bobbobbio commented on src/url.rs:
```diff
@@ -37,6 +40,86 @@ impl fmt::Display for Scheme {
+    pub fn host(&self) -> &str {
+        &self.host
+    }
+    // give use an option to use the origin url struct and save us from porting too many codes
```
I wasn't able to understand what this comment was trying to say

ivila replied:
Oh, sorry for my poor english, I was trying to say that giving the user of http_io an option to use the origin url struct, so we won't need to imply other functions(like path, username, password, etc)

I will delete this comment as it was misleading.

bobbobbio commented on src/url.rs:
```diff
@@ -37,6 +40,86 @@ impl fmt::Display for Scheme {
+        use core::str::FromStr;
+
+        let scheme = Scheme::from_str(url.scheme())?;
+        if scheme.ne(&Scheme::Http) && scheme.ne(&Scheme::Https) {
```
I think you can use the `!=` operator instead of the explicit function call

bobbobbio commented on src/url.rs:
```diff
@@ -37,6 +40,86 @@ impl fmt::Display for Scheme {
+    pub fn port(&self) -> Result<u16> {
+        match self.url.port_or_known_default() {
+            Some(v) => Ok(v),
+            None => Err(Error::UrlError(format!(
```
Can you write a test that exercises this error path? As it stands I wonder if it is ever hit, since we now always know the scheme is HTTP or HTTPS

ivila replied:
the port must goes to happy path, I will change it to just unwrap and left comments with it

bobbobbio commented on src/url.rs:
```diff
@@ -37,6 +40,86 @@ impl fmt::Display for Scheme {
+impl TryFrom<Url> for HttpUrl {
+    type Error = Error;
+
+    fn try_from(url: Url) -> Result<Self> {
```
I think this function deserves some direct tests in this file

bobbobbio commented on src/url.rs:
```diff
@@ -221,4 +304,20 @@ mod tests {
+
+    #[test]
+    fn parse_http_url() {
+        // test parse http url from str, mainly watch the scheme, host and port
```
I would delete this comment, in my opinion it doesn't make anything more clear

bobbobbio commented on src/ssl/mod.rs:
```diff
@@ -1,4 +1,4 @@
-#[allow(unused)]
+#[allow(dead_code)] // disable warning of not using the String in the Error
```
this comment is good, but I feel it is missing an explanation of why this is necessary (instead of just removing String), I bet conditional compilation is the reason.

ivila replied:
This is some kind of false positive of cargo linter, it report that the String is not used by the Error struct, and it notice that you have derive Debug(and this is why we cannot remove the String, because if we remove it, then users of http_io cannot know the description of the error anymore)

Here is the error message, the linter knows you are using it, but it just will report this.
![image](https://github.com/user-attachments/assets/eaca14cb-5a04-4e0c-9cca-7b352a643f9d)

bobbobbio commented on src/url.rs:
```diff
@@ -37,6 +40,86 @@ impl fmt::Display for Scheme {
+            )));
+        };
+        let Some(host) = url.host_str() else {
+            // people can use url.set_host(Option::None) to make host invalid, and currently we
```
that is a good point to make, but also I think some URLs just don't have a host from the get-go, so that sounds like the two situations this is covering and as you have the comment it feels like you should mention the other one

ivila replied:
after testing, I think the url crate have ensure that there must have host in http and https.
I will change it to just unwrap, and add a test for declaration

bobbobbio reviewed (changes requested):
Thank you so much for implementing my feedback.

This is very much improved and think it is very close. The main two things I want you look at before I approve it is better test coverage for the new `try_from` implementation you wrote, and I want you too look at if `port()` really needs to return an error.

ivila committed:
PR fix:
1) use != instead of ne
2) use certain result of host and port as we know the scheme must be http or https
3) add some tests

ivila committed:
add ci file

ivila commented:
I also add a ci, see https://github.com/ivila/http_io/actions/runs/11905527717

bobbobbio commented on src/url.rs:
```diff
@@ -48,48 +48,38 @@ pub struct HttpUrl {
-            ))),
-        }
+    pub fn port(&self) -> u16 {
+        // this should always go to the happy path as we only have scheme of http and https
```
I would phrase this as "this will never fail because we verified the scheme is HTTP or HTTPS which should always have a port"

bobbobbio commented on src/url.rs:
```diff
@@ -48,48 +48,38 @@ pub struct HttpUrl {
+        if scheme != Scheme::Http && scheme != Scheme::Https {
+            return Err(error_unsupported_url_scheme(url.scheme()));
         };
+        // host must exist in http and https, and url crate ensure about it, see test check_url_must_have_host
```
I would phrase this as "HTTP and HTTPS URLs must always have a host, see the check_url_must_have_host test"

bobbobbio reviewed (approved):
With the new changes, it looks great. Thanks so much.

ivila committed:
PR fix: fix comments
